### PR TITLE
NXP IMXRT API to use half-duplex communication with serial driver.

### PR DESCRIPTION
**Pull request by Hellspread41**
**Added Half-Duplex Support for UART on NXP i.MX RT**

This update introduces two new functions to the UART driver for **NXP i.MX RT family microcontrollers**, enabling Half-Duplex (master-slave) communication with internal driver components. The added functions enhance the UART driver's flexibility by offering the capability to configure Half-Duplex communications, crucial for applications requiring master-slave setups, thus reducing resource usage and improving communication efficiency.

Changes includes:
- A new Kconfig option for uart driver called `CONFIG_UART_BUILTIN_HALFDUPLEX` to activate the following API.
- The function: `int uart_hd_talk()` allowing the master to send a packet in half-duplex mode.
- The function: `int uart_hd_listen()` allowing the master to listen to incoming packet from slave sharing the same wire.
- A new configuration inside the binding file _nxp,mcux_lpuart.yaml_ to enable this mode inside the devicetree.

These additions facilitate the integration and communication with other internal components in Half-Duplex mode, thereby offering greater versatility in embedded system designs using NXP i.MX RT microcontrollers.

Developed for a personal project that required this type of communication to control Dynamixel smart servos directly without using a converter.

Simple exemple using polling api:
```c
#include <zephyr/logging/log.h>
LOG_MODULE_REGISTER(hd_test, LOG_LEVEL_DBG);

#include <zephyr/kernel.h>
#include <zephyr/drivers/uart.h>

/* lpuart3 device structure. */
const struct device* lpuart3_dev = DEVICE_DT_GET(DT_NODELABEL(lpuart3));
const uint8_t ping_pkt[] = {0xFF, 0xFF, 0xFD, 0x00, 0x01, 0x03, 0x00, 0x01, 0x19, 0x4E};


static void lpuart3_cb(const struct device *dev, void *user_data) 
{
    ARG_UNUSED(user_data);
    uint8_t byte_buf;

    if((uart_irq_update(dev) > 0) && (uart_irq_is_pending(dev) > 0)) {
        if(uart_irq_rx_ready(dev)) {
            int err = uart_fifo_read(dev, &byte_buf, sizeof(uint8_t));
            if(err > 0) LOG_INF("rx byte = 0x%x", byte_buf);
        }
    }
}

static int halfduplex_uart_init(const struct device* uart_dev, uint32_t baudrate, uart_irq_callback_user_data_t isr) {

    int err;

    if(!device_is_ready(uart_dev)) {
        LOG_ERR("uart device is not ready to be used");
        return -1;
    }

    const struct uart_config config = {
        .baudrate = baudrate,
        .data_bits = UART_CFG_DATA_BITS_8,
        .flow_ctrl = UART_CFG_FLOW_CTRL_NONE,
        .parity = UART_CFG_PARITY_NONE,
        .stop_bits = UART_CFG_STOP_BITS_1
    };

    err = uart_configure(uart_dev, &config);
    if(err != 0) {
        LOG_ERR("something went wrong during the configuration of the uart device (err: %d)", err);
        return err;
    }

    err = uart_irq_callback_user_data_set(uart_dev, isr, NULL);
    if(err != 0) {
        LOG_ERR("something went wrong during callback attribution for interrupt driven functionnality (er; %d)", err);
        return err;
    }

    uart_irq_tx_disable(uart_dev);
    uart_irq_rx_enable(uart_dev);

    return 0;
}

int main(void) {
    LOG_INF("application entry point");
    int err;

    err = halfduplex_uart_init(lpuart3_dev, 1000000, lpuart3_cb);
    if(err != 0) {
        LOG_ERR("uart configuration failed (err: %d)", err);
        return -1;
    }

    LOG_INF("halfduplex configuration success");

    k_sleep(K_SECONDS(1));
    
   /* send a packet the the dynamixel servo */
    err = uart_hd_talk(lpuart3_dev);
    for(uint8_t i = 0; i < sizeof(ping_pkt); i++) {
        uart_poll_out(lpuart3_dev, ping_pkt[i]);
    }
   /* put master into listening mode to receive the status packet from the slave */
    err = uart_hd_listen(lpuart3_dev);

    return 0;
}


```

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the
    best of my knowledge, is covered under an appropriate open
    source license and I have the right under that license to
    submit that work with modifications, whether created in whole
    or in part by me, under the same open source license (unless
    I am permitted to submit under a different license), as
    Indicated in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including
    all personal information I submit with it, including my
    sign-off) is maintained indefinitely and may be redistributed
    consistent with this project or the open source license(s)
    involved.

Signed-off-by: Simon Marceau <marceau41@hotmail.be>

### CODE REVIEW

File: drivers/serial/Kconfig
Comment (pdgendt) [inline drivers/serial/Kconfig:161]: `default n` is the default, omit this line.

**Review by pdgendt — CHANGES_REQUESTED** [review]
Indentation should use tabs instead of spaces.

File: drivers/serial/Kconfig
Comment (Hellspread41) [reply drivers/serial/Kconfig:161]: did not notice this thanks

**Comment by dcpleung** [comment]
Couple things:
* This also needs `CONFIG_SERIAL_HAS_HALFDUPLEX` to indicate whether a driver support the half duplex API.
* The kconfig should simply be `CONFIG_SERIAL_HALFDUPLEX_API`, and skip the "built-in" part.
* The naming of APIs should not be simply `*_hd_*`. They need to be more verbose.
* Need to separate the API changes and driver changes into different commits.

**Review by dcpleung — CHANGES_REQUESTED** [review]
(see above)

**Review by dcpleung — CHANGES_REQUESTED** [review]
Please do not dismiss change requests on your own and let the reviewer do it.

**Review by decsny — CHANGES_REQUESTED** [review]
- how will this be tested in continuous integration?
- how does being in "listen mode" or "talking mode" effect the rest of the api usage
- it seems like the devicetree property should be generic
- perhaps the duplex should be set in uart_config api instead of devicetree 
- is there a reason why there cant be one function to switch half duplex direction?

**Comment by Hellspread41** [comment]
Everything was tested with the __mimxrt1050_evk_ board and the slave was a servomotor from dynamixel company using halfuplex communication.

I've tested with polling, interrupt and async api and was working fine. If the driver is not in listen mode during a packet receiving, we receive nothing.
So it is the responibility of the user to change the mode at the right moment.

I've been inspired by the loopback_en devicetree property to devellop the halfduplex api.

**Comment by decsny** [comment]
> Everything was tested with the __mimxrt1050_evk_ board and the slave was a servomotor from dynamixel company using halfuplex communication.

What I mean is, this api needs to be tested in the uart test cases 

> I've tested with polling, interrupt and async api and was working fine. If the driver is not in listen mode during a packet receiving, we receive nothing. So it is the responibility of the user to change the mode at the right moment.

The exact behavior and implications should be documented in the API documentation

> I've been inspired by the loopback_en devicetree property to devellop the halfduplex api.

Did not answer my other questions

**Comment by Hellspread41** [comment]
> What I mean is, this api needs to be tested in the uart test cases

No problem! I'm new to using the Zephyr framework for unit testing, so I still need to figure out how to manage it effectively.

> The exact behavior and implications should be documented in the API documentation.

I agree, I need to provide more precise comments on the APIs regarding the consequences of misusing them.

> it seems like the devicetree property should be generic

What do you mean with "generic" ?

> perhaps the duplex should be set in uart_config api instead of devicetree

To my knowledge, this specificity is only present in this NXP driver. Therefore, I believed it was best to include it as a devicetree property. Would placing this option in the uart_config structure make sense if other manufacturers offer this feature ?

> is there a reason why there cant be one function to switch half duplex direction?

I see, there isn't really a reason for that. It is true that the direction of communication could be reversed by using only one function that takes the direction as a parameter. I prefer to have the opinion of more experienced developers.

Thank you for your valuable insights and expertise.

**Comment by decsny** [comment]
> > What I mean is, this api needs to be tested in the uart test cases
> 
> No problem! I'm new to using the Zephyr framework for unit testing, so I still need to figure out how to manage it effectively.

https://docs.zephyrproject.org/latest/develop/test/ztest.html

> > The exact behavior and implications should be documented in the API documentation.
> 
> I agree, I need to provide more precise comments on the APIs regarding the consequences of misusing them.
> 
> > it seems like the devicetree property should be generic
> 
> What do you mean with "generic" ?

I mean perhaps this property or one like it could be used by any uart in zephyr instead of just the one NXP driver

> > perhaps the duplex should be set in uart_config api instead of devicetree
> 
> To my knowledge, this specificity is only present in this NXP driver. Therefore, I believed it was best to include it as a devicetree property. Would placing this option in the uart_config structure make sense if other manufacturers offer this feature ?

I am not sure, this is only an idea, not necessarily a good one, I'll let uart maintainer comment

**Pull request closed without merging**
